Improve onSave() processing time and Fix all parts getting marked as tested on load

tathxo opened the pull request:
### Improve onSave() processing time
The onSave() function in Data.cs calls .Contains() function on the testedParts list for every part. list.Contains() is linear (scales to number of entries) and when most parts are in testedParts, this ends up being a function with O(n^2) time complexity.

At 400,000 parts, this results in saves taking ~50 seconds each, which is a problem not only for saving + autosaves, but also at any point in the game when KSP saves to the "persistent" save file (situation change, etc).

Making the testedParts list instead a HashSet makes the Contains() function roughly O(1) time complexity, which makes the whole onSave() function only O(n) time complexity - or linearly to the number of parts. In the 400,000+ parts case, this reduces the save time significantly, down to a few seconds.

----

### Fix all parts getting marked as tested on load
TryParse() returns true whether the "Tested" value is true or false. Since this is used for populating the "Tested Parts", loading the game ends up marking every part as tested (even if not actually tested). If a player loads often, this results in every part being marked as tested (and a huge number of tested parts).

By checking the output boolean b instead, it will keep the "Tested" value instead.

tathxo committed:
Improve onSave() processing time

The onSave() function in Data.cs calls .Contains() function on the testedParts list for every part. list.Contains() is linear (scales to number of entries) and when most parts are in testedParts, this ends up being a function with O(n^2) time complexity.

At 400,000 parts, this results in saves taking ~50 seconds each, which is a problem not only for saving + autosaves, but also at any point in the game when KSP saves to the "persistent" save file (situation change, etc).

Making the testedParts list instead a HashSet makes the Contains() function roughly O(1) time complexity, which makes the whole onSave() function only O(n) time complexity - or linearly to the number of parts. In the 400,000+ parts case, this reduces the save time significantly, down to a few seconds.

tathxo committed:
Fix all parts getting marked as tested on load

TryParse() returns true whether the "Tested" value is true or false. Since this is used for populating the "Tested Parts", loading the game ends up marking every part as tested (even if not actually tested). If a player loads often, this results in every part being marked as tested (and a huge number of tested parts).

By checking the output boolean b instead, it will keep the "Tested" value instead.

tathxo committed:
Fix build issue

This fixes an issue where Visual Studio complains about the unpaired Reference end tag

Bobisback commented:
How hard is it to get a new build with this? Guessing I need to pull down the repo in your branch and build locally to get these changes?

P.S It looks like this should target the https://github.com/zer0Kerbal/OhScrap/tree/2.2.99.1-prerelease branch as that appears to be the latest branch, not master.

I am going to put your changes in and then rebuild using this branch.

tathxo commented:
@Bobisback It's been 7.5 months since I made this request and about 6 months since my last dive into KSP ended so I'm a bit hazy on the details.

I'm not really familiar with git best practices, so if it should go to prerelease branch that's completely fine. Either you can put it in on that or if desired I can move it over (and I can close this request either way). I don't really mind about getting credit for authorship or whatever so feel free to commit it yourself. Most of the work is just in getting it to build, and testing it in KSP with an appropriately big save.

Regarding the build, similarly to git I'm not familiar with Visual Studio best practices, but I was able to build it locally after:
1) Resolving target framework issues (targeting .NET 4.7.2 on VS Community 2022 v17.11.2)
2) Making some modification to/removing the .csproj's references, 
3) Adding new References to resolve each of the sets of errors that came up upon build, and 
4) Removing the post-build events entirely, since they were causing issues (not sure _addonDeploy.bat or _addonRelease.bat are even present in the repo to begin with).

In hindsight, instead of doing 2+3 where I deleted the csproj references then re-added them, I probably could have just tried to assign the missing/undefined environment variables (K100 or K112?), as well as moved its referenced code/objects into the right folder structures.
As a result of this realization I only included a fix to the .csproj for a doubly-closed tag which was giving me problems - and that issue does not appear to be in the file on that pre-release branch.

To test with the resulting build, I believe I just dropped the built .dll over where CKAN had put its OhScrap .dll into my KSP install when I had installed it through CKAN. As I recall I had to also copy some localization files in to fix some in-game text, because the latest CKAN release didn't have any of that.
